File: App-13/app/src/main/java/com/android_lab/app_13/MyAdapter.java
package com.android_lab.app_13;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;


public class MyAdapter extends RecyclerView.Adapter<MyAdapter.MyViewHolder> {
    public ArrayList<Message> dataset;
    private ClickListener mListener;

    public static interface ClickListener {
        public void onClick(int position);
    }

    public static class MyViewHolder extends RecyclerView.ViewHolder {

        public TextView title, subtitle;
        public ClickListener mListener;

        public MyViewHolder(View v, ClickListener mListener) {
            super(v);
            title = v.findViewById(R.id.title);
            subtitle = v.findViewById(R.id.subtitle);

            this.mListener = mListener;
            v.setOnClickListener(onClickListener);
        }

        private final View.OnClickListener onClickListener = new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                mListener.onClick(getAdapterPosition());
            }
        };
    }

    public MyAdapter(ArrayList<Message> dataset, ClickListener mListener) {
        this.dataset = dataset;
        this.mListener = mListener;
    }

    @Override
    public MyAdapter.MyViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View v = (View) LayoutInflater.from(parent.getContext()).inflate(R.layout.cell_layout, parent, false);

        MyViewHolder vh = new MyViewHolder(v, mListener);
        return vh;
    }

    @Override
    public void onBindViewHolder(MyViewHolder holder, int position) {
        holder.title.setText(dataset.get(position).address);
        holder.subtitle.setText(dataset.get(position).body);
    }

    @Override
    public int getItemCount() {
        return dataset.size();
    }
}

File: App-11_kotlin/app/src/main/java/com/android_lab/app_11_kotlin/MyAdapter.kt
package com.android_lab.app_11_kotlin

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView

class MyAdapter(private val dataset: Array<Car>, private val mListener: ClickListener): RecyclerView.Adapter<MyAdapter.MyViewHolder>() {

    interface ClickListener {
        fun onClick(position: Int)
    }

    class MyViewHolder(val view: View, val mListener: ClickListener): RecyclerView.ViewHolder(view) {
        var title: TextView
        val onClickListener = View.OnClickListener {
            mListener.onClick(adapterPosition)
        }

        init {
            title = view.findViewById(R.id.title)
            view.setOnClickListener(onClickListener)
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MyViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.cell_layout, parent, false)

        return MyViewHolder(view, mListener)
    }

    override fun onBindViewHolder(holder: MyViewHolder, position: Int) {
        holder.title.text = dataset[position].name
    }

    override fun getItemCount() = dataset.size

}